### ISSUE TITLE
IMap.keySet() fetches all values over the network

### ISSUE THREAD

**Issue by lukasblu** [hazelcast/hazelcast#5634]
Hi,

in our application, besides several other IMaps, we also use a IMap to keep PDFs in main memory - so the size of the values in this map are quite large (e.g. 250 kB).

While testing the performance of our application, I realized that a call to IMap.keySet() not only fetches the keys, but also the values! This seriously affects the network load, making our performance tests fail misserably...

The problem did not exist in Hazelcast version 3.3, but was introduced in version 3.4 and still exists in version 3.5. Please fix this as soon as possible and if anyhow possible, please also backport to version 3.4.

Measuring network traffic in a JUnit test is kind of difficult, but if you are using Linux and have iptables installed, you can check the traffic on the loopback device. The following test does this, but needs to be run as root (as iptables can only be executed as root on my machine).

```java
package com.nm.test.hazelcast.map;

import com.hazelcast.config.Config;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import com.nm.test.hazelcast.TestHazelcast;
import com.nm.test.hazelcast.utils.StartHcInstances;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.Serializable;
import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.List;
import java.util.Random;
import java.util.Set;
import junit.framework.TestCase;

/**
 * A test to ensure IMap.keySet() only fetches the keys - and not also the values over the network.
 */
public class TestMap2 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestMap2.class);

	private static final String mapName = "testMap" + TestMap2.class.getSimpleName();

	public static void main(String[] args) throws Exception {
		TestMap2 test = new TestMap2();
		test.setUp();
		test.testKeySet();
	}

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testKeySet() throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// use initial min cluster size
		config.setProperty("hazelcast.initial.min.cluster.size", "2");

		// create two Hazelcast instances
		List<HazelcastInstance> hcInstances = new StartHcInstances(config, 2).getHcInstances();
		HazelcastInstance hcInstance1 = hcInstances.get(0);
		logger.info("Hazelcast cluster created. Cluster size = " + hcInstance1.getCluster().getMembers().size());

		// try-finally to stop hazelcast instances
		try {

			// put entries into map
			int numKeys = 1000;
			NumberFormat format = new DecimalFormat("00000");
			IMap<String, Blob> map = hcInstance1.getMap(mapName);
			for (int i = 0; i < numKeys; i++) {
				String key = "key__" + format.format(i);
				map.put(key, new Blob());
			}
			logger.info("Map size after put: " + map.size());

			// zero loopback traffic
			zeroLoopbackTraffic();
			long traffic0 = getLoopbackTraffic();
			logger.info("Traffic before in bytes: " + traffic0);

			// call keySet multiple times
			int repeatTimes = 20;
			for (int i = 0; i < repeatTimes; i++) {
				Set<String> keys = map.keySet();
				StringBuilder buf = new StringBuilder();
				for (String key : keys) {
					buf.append(key);
				}
				assertTrue("Unexpected total key length!", buf.length() == numKeys * 10);
			}

			// measure traffic
			long traffic1 = getLoopbackTraffic();
			long trafficTotal = traffic1 - traffic0;
			logger.info("Traffic after in bytes: " + traffic1);
			logger.info("Measured traffic in bytes: " + trafficTotal);

			// fail if too much traffic
			long trafficMax = repeatTimes * (numKeys * 10 * 5);
			logger.info("Max expected traffic in bytes: " + trafficMax);
			logger.info("Factor: " + ((double) trafficTotal / trafficMax));
			assertTrue("Too much traffic!", trafficTotal < trafficMax);

		} finally {
			for (HazelcastInstance hcInstanceCur : hcInstances) {
				hcInstanceCur.getLifecycleService().terminate();
			}
		}

		logger.info("Test done.");
	}

	public static class Blob implements Serializable {

		private static final long serialVersionUID = 1L;

		private byte[] data;

		public Blob() {
			data = new byte[250000];
			Random random = new Random();
			random.nextBytes(data);
		}
	}

	private void zeroLoopbackTraffic() throws Exception {

		// iptables --zero
		String cmd = "iptables --zero";
		logger.info("executing: '" + cmd + "'...");
		Process p = Runtime.getRuntime().exec(cmd);
		BufferedReader input = new BufferedReader(new InputStreamReader(p.getInputStream()));
		BufferedReader error = new BufferedReader(new InputStreamReader(p.getErrorStream()));
		String line = null;
		logger.debug("input:");
		while ((line = input.readLine()) != null) {
			logger.debug(line);
		}
		logger.debug("error:");
		while ((line = error.readLine()) != null) {
			logger.debug(line);
			if (line.contains("Permission denied (you must be root)")) {
				throw new Exception(line);
			}
		}
	}

	private long getLoopbackTraffic() throws Exception {

		// iptables -vnL
		// 21579  236M ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
		long bytes = -1;
		String cmd = "iptables -vnL";
		logger.info("executing: '" + cmd + "'...");
		Process p = Runtime.getRuntime().exec(cmd);
		BufferedReader input = new BufferedReader(new InputStreamReader(p.getInputStream()));
		BufferedReader error = new BufferedReader(new InputStreamReader(p.getErrorStream()));
		logger.debug("input:");
		String line = null;
		while ((line = input.readLine()) != null) {
			logger.debug(line);
			if (bytes == -1 && line.contains(" lo ")) {
				bytes = parseTraffic(line);
			}
		}
		logger.debug("error:");
		while ((line = error.readLine()) != null) {
			logger.debug(line);
			if (line.contains("Permission denied (you must be root)")) {
				throw new Exception(line);
			}
		}

		return bytes;
	}

	private long parseTraffic(String line) {
		String[] cols = line.trim().split("\\s+");
		String trafficStr = cols[1];
		System.out.println("Traffic: " + cols[1]);
		long bytes = 0;
		if (trafficStr.endsWith("K")) {
			bytes = Long.parseLong(trafficStr.substring(0, trafficStr.length() - 1)) * 1024;
		} else if (trafficStr.endsWith("M")) {
			bytes = Long.parseLong(trafficStr.substring(0, trafficStr.length() - 1)) * 1024 * 1024;
		} else if (trafficStr.endsWith("G")) {
			bytes = Long.parseLong(trafficStr.substring(0, trafficStr.length() - 1)) * 1024 * 1024 * 1024;
		} else {
			bytes = Long.parseLong(trafficStr);
		}
		return bytes;
	}

}
```

Thanks and best,
Lukas

**Comment by Donnerbart** [comment]
Duplicate of https://github.com/hazelcast/hazelcast/issues/4642

**Comment by Donnerbart** [comment]
It's also the same for IMap.values(), basically everything which is using a Predicate internally. The whole entry set is transported to the requesting node and that one filters the key, value or copies the whole entry.

Maybe it was changed in 3.4 to do use Predicates instead of a handcrafted operation. The Java client e.g. is still using it's own methods, so it should not be affected.

I think the fix for this should be to make the queries smarter (and pass the IterationType to all members to filter directly), instead of going back to unique operations for each method. In this way also the queries will take advantage from the fix.

**Comment by Donnerbart** [comment]
This should be the scope of affected methods:
![Scope](http://docs.hazelcast.org/docs/3.5/manual/html-single/images/Map-QueryResultSizeLimiterScope.png)

I didn't refactor the client to use queries as well due to this issue.

**Comment by lukasblu** [comment]
Hi @Donnerbart ,

sorry for missing the issue #4642 . I was shortly looking at the list of open issues, but was a bit overwhelmed by the sheer number of open ones :-) next time I will try to search for 'keySet' or something...

Anyway, thanks for keeping this issue open due to its test code. To make the test code complete, here is the missing class StartHcInstances:

```java
package com.nm.test.hazelcast.utils;

import com.hazelcast.config.Config;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import java.util.ArrayList;
import java.util.List;

public class StartHcInstances {

	private List<HazelcastInstance> hcInstances = new ArrayList<HazelcastInstance>();

	public StartHcInstances(Config config, final int numNodes) throws Exception {

		// create runnables and threads
		List<HcStartRunnable> runnables = new ArrayList<StartHcInstances.HcStartRunnable>();
		List<Thread> threads = new ArrayList<Thread>();
		for (int i = 0; i < numNodes; i++) {
			HcStartRunnable runnable = new HcStartRunnable(config);
			runnables.add(runnable);
			threads.add(new Thread(runnable, "node" + i));
		}

		// start and wait for completion
		for (Thread thread : threads) {
			thread.start();
		}
		for (Thread thread : threads) {
			thread.join();
		}

		// collect running hazelcast instances
		for (HcStartRunnable runnable : runnables) {
			hcInstances.add(runnable.getHcInstance());
		}
	}

	public List<HazelcastInstance> getHcInstances() {
		return hcInstances;
	}

	private static class HcStartRunnable implements Runnable {

		private final Config config;

		private HazelcastInstance hcInstance;

		public HcStartRunnable(Config config) {
			this.config = config;
		}

		@Override
		public void run() {
			hcInstance = Hazelcast.newHazelcastInstance(config);
		}

		public HazelcastInstance getHcInstance() {
			return hcInstance;
		}

	}

}
```

If running this test stand alone, you can change the setup method as follows to get rid of the dependency on TestHazelcast:

```java

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}
```
change to
```java

	@Override
	protected void setUp() throws Exception {
		BasicConfigurator.configure();
	}
```

thanks and best,
Lukas

Ps. I see you added this to the 3.6 milestone - are you really not planning to fix this in 3.4 and/or 3.5?

**Comment by Donnerbart** [comment]
It's the next milestone which is open and I (personally) want this fixed, so I put it on there. We'll see if there is time for this in the 3.6 schedule.

We also have to start fixing in the master branch, so that's 3.6. If it's done we'll see if we can backport the changes to older versions, if that's possible.

**Comment by kraythen** [comment]
We have been affected by this as well. The workaround we employed was to do submitToAllMembers and then  run the predicate against the localKeySet and then return the results to the caller. 

I would classify this as a MAJOR issue. Resulted in traffic between our nodes in the terabyte range.

**Comment by jerrinot** [comment]
related experimental PR: #6215

**Comment by jerrinot** [comment]
This is fixed by #6288

**Comment by andrey-yemelyanov** [comment]
Hi,

I am seeing this problem in Hazelcast 4.1.1. When calling map.keySet(), values will be deserialized. Is it a feature now? Please advise how I can retrieve just the keys in a certain map.
